### ISSUE TITLE
Use repair to rebuild a tablet replica 

### ISSUE THREAD

**Issue by tgrabiec** [scylladb/scylladb#17174]
Currently, when rebuilding a new tablet replica (e.g. after removenode) we stream data from all the remaining replicas. That's not ideal because:
  1) typically all replicas contain mostly the same data, so there is a lot of redundancy. We waste bandwidth and add work for compaction which burns disk and CPU resources  
  3) if streaming is faster than compaction, we bloat the disk usage on the receiving node. This is not a problem for individual tablet since it contains only a fraction of data on a node. But if lots of such rebuilds happen on the node, the effect compounds and there is no backpressure from compaction. Avi suggested to add feedback loop between disk usage and tablet scheduler to avoid this problem.    

We should instead:
  1) repair remaining replicas
  2) stream from one of the replicas

Note that the repair in step 1 is partial, so it cannot mark the tablet as repaired (set gc threshold). It can do so only after the streaming step completes.

**Comment by mykaul** [comment]
Priority? 6.0 material?

**Comment by bhalevy** [comment]
@tgrabiec / @raphaelsc: in the current implementation, do we / should we trigger major compaction after streaming, or is regular compaction good enough? (and maybe the min_threshold can be lowered temporarily to ensure compaction kicks-in)

**Comment by bhalevy** [comment]
@raphaelsc ping, please see question above.

**Comment by raphaelsc** [comment]
> @tgrabiec / @raphaelsc: in the current implementation, do we / should we trigger major compaction after streaming, or is regular compaction good enough? (and maybe the min_threshold can be lowered temporarily to ensure compaction kicks-in)

regular compaction should be able to pick it. but note we still need the solution presented (== stream from one of the replicas) to avoid disk space issues, as streaming can produce data faster than compaction can consume it.

**Comment by raphaelsc** [comment]
should we consider this a release blocker?

**Comment by raphaelsc** [comment]
the problem is amplified significantly with file based streaming.

**Comment by bhalevy** [comment]
> should we consider this a release blocker?

No, since it's an improvement

**Comment by bhalevy** [comment]
> We should instead:
> 
> 1. repair remaining replicas
> 2. stream from one of the replicas
> 
> Note that the repair in step 1 is partial, so it cannot mark the tablet as repaired (set gc threshold). It can do so only after the streaming step completes.

@tgrabiec if we knew which of the tablet replicas needs rebuild, we could reverse the steps:
1. stream from any of the non-pending replicas to the the pending replicas
2. repair replica set

Then the repair step could be final and we could mark the tablet as repaired.

**Comment by denesb** [comment]
@bhalevy @xemul @tgrabiec is this issue still valid? Do we still want to use repair to rebuild? I lost track on this after all the discussion around adding DCs.

**Comment by bhalevy** [comment]
> This has clearly missed 6.0, unless it was implemented ... ?

No, it's an optimization that can,be deferred.
Certainly after we have native tablet repair that is intrgrated into the tablet scheduler.
We should not harness repair for rebuild otherwise, while it loosly coupled with tablet transitions.

**Comment by mykaul** [comment]
> > This has clearly missed 6.0, unless it was implemented ... ?
> 
> No, it's an optimization that can,be deferred. Certainly after we have native tablet repair that is intrgrated into the tablet scheduler. We should not harness repair for rebuild otherwise, while it loosly coupled with tablet transitions.

@bhalevy - do we need this now?

**Comment by bhalevy** [comment]
> > > This has clearly missed 6.0, unless it was implemented ... ?
> > 
> > 
> > No, it's an optimization that can,be deferred. Certainly after we have native tablet repair that is intrgrated into the tablet scheduler. We should not harness repair for rebuild otherwise, while it loosly coupled with tablet transitions.
> 
> [@bhalevy](https://github.com/bhalevy) - do we need this now?

It will be good to integrate rebuild with native tablet repair, not that we have it.

That said, we may want to distinguish between two use cases of running `nodetool rebuild`:
1. After losing local data. In this case, it makes sense to run repair to rebuild the missing data for all keyspaces, vs.
2. Running `nodetool rebuild` as part of adding a new DC, after increasing the replication factor.

In the latter case, `nodetool rebuild` is part of the documented procedure and is required for rebuilding replicas in vnode keyspaces, not for tablet keyspaces (their replicas are rebuilt automatically as part of increasing the replication far tor).

We can add a `--repair` (or `--all-keyspaces`) option to `nodetool rebuild` to enable repair for the tablet keyspaces,
and mention it in the warning we already print when running `nodetool rebuild` when there are tablet keyspaces in the system.

**Comment by bhalevy** [comment]
> Currently, when rebuilding a new tablet replica (e.g. after removenode) we stream data from all the remaining replicas. That's not ideal because:
> 
> 1. typically all replicas contain mostly the same data, so there is a lot of redundancy. We waste bandwidth and add work for compaction which burns disk and CPU resources
> 2. if streaming is faster than compaction, we bloat the disk usage on the receiving node. This is not a problem for individual tablet since it contains only a fraction of data on a node. But if lots of such rebuilds happen on the node, the effect compounds and there is no backpressure from compaction. Avi suggested to add feedback loop between disk usage and tablet scheduler to avoid this problem.
> 
> We should instead:
> 
> 1. repair remaining replicas
> 2. stream from one of the replicas
> 
> Note that the repair in step 1 is partial, so it cannot mark the tablet as repaired (set gc threshold). It can do so only after the streaming step completes.

Note that the proposal would work well with:
- Incremental repair:
  - The dataset to be repaired would be small
  - Assuming inc. repair starts after double writes (write both, read old), streaming can be done only from the repaired sstable-set
    - Since the writes that happened after repair started would be in the unrepaired sstable set
- 0->N RF change
  -  In this case after repair we can stream in parallel by pairing the source-dc replicas to the new dc replicas
  - Note: rf would typically be the same in the source and destination DCs, but even if not, we can stream from N->M

Disadvantages of existing solution based on file-streaming from all existing replicas to the new replicas:
- It consumes a lot of inter-region bandwidth
  - which is limited and expensive

**Comment by bhalevy** [comment]
Also, upped priority from P2 to P1 since the existing solution causes high space amplification when adding a new datacenter (e.g. 3x for the 2nd DC, or 6x when adding a 3rd DC)

**Comment by denesb** [comment]
I reassigned to @Deexie, she can get started after finishing the tablet repair follow-up.
That said, I don't know how realistic 2025.1.x milestone is.

**Comment by dani-tweig** [comment]
> I reassigned to [@Deexie](https://github.com/Deexie), she can get started after finishing the tablet repair follow-up. That said, I don't know how realistic 2025.1.x milestone is.

@denesb will it make to 2025.2.0? If not, please adjust the milestone according to your assessments

**Comment by denesb** [comment]
Pushed out to 2025.3.
